Update Travis configuration to also build on oraclejdk9

TimvdLippe opened the pull request:
Let's see if this actually works. Testing for #520

TimvdLippe commented:
All right, for this one the `HTMLDoclet` is actually breaking the build, which was apparently the issue in #521. Therefore reopening that one. JDK8 is working, but it seems they have moved some classes in JDK9. Not sure if we can have a configuration which works for both JDKs.

TimvdLippe commented:
I have started the configuration to only compile certain files on a specific JDK version, to at least not fail compilation on JDK 9. However as I personally do not have JDK 9 installed (still running on JDK 8) I can't develop a valid doclet. Does anyone else want to take over implementing the exclude function?

bric3 commented on gradle/javadoc.gradle:
```diff
@@ -14,7 +14,7 @@ task mockitoJavadoc(type: Javadoc) {
     def javadocJdk6Hack = "jdk6-project-version-insert.min.js"
 
-    options.doclet "org.mockito.javadoc.JavadocExclude"
+    options.doclet "org.mockito.javadoc.JavadocExclude1_${JavaVersion.current().toString()[-1]}"
```
It is a one-liner yet I find the name ugly ;)
Also this won't work with Java 10.

bric3 commented:
Otherwise it LGTM

TimvdLippe commented:
**One eternity later**

All right this took quite a while, but we can actually compile on JDK9 now :tada:. This required some "workarounds" to bugs in various versions of our dependencies.

* I updated Gradle to 3.1. This release fixes an internal ClassCastException when compiling Groovy buildSrc code
* I updated Jacoco to the latest snapshot release to get around several exceptions thrown in the bytecode generated by Jacoco. The corresponding PR is https://github.com/jacoco/jacoco/pull/434

I sadly had to remove the doclet for JDK > 8, because the latest release available for xenial release is build 114 which does not expose the actual doclets and internal packages required to write a doclet. This is fixed in later builds, but these are only published for the development version of ubuntu :cry: 

All in all we are now compiling correctly, but sadly we have failing tests: https://travis-ci.org/mockito/mockito/jobs/161332991#L433-L465 These seem to be related to AssertJ failing to understand generics.

I have also seen other test failures, so maybe some other tests are failing still. @szpak could you try this branch and see if it compiles locally for you?

TimvdLippe commented:
Only 1 more failure with the inline mock maker tests: https://travis-ci.org/mockito/mockito/jobs/164437417#L396-L401

raphw commented:
ASM 6 (support for Java 9) is still in an alpha state, therefore Java 9 class files are still rejected. The moment ASM 6 is released, this is updated.

raphw commented:
Optionally, we can use byte-buddy-dep and explicitly use ASM 6 alpha.

bric3 commented:
I think we should keep the actual bytebuddy. We could eventually disable the test with JDK 9 using `assumeThat`

raphw commented:
I do not think that Remi will release ASM 6 before the class file format is finalized and I do not think it is a good idea to rely on these things yet. Without the inlining, Byte Buddy does not need to process any Java 9 class files which is why I would suggest to add a rule to the inliner tests to not be run on Java 9 (yet).

TimvdLippe commented:
I will disable the test on jdk9 then, thanks for the explanation!

TimvdLippe commented:
Ignoring the broken test on JDK9. This PR should now be green. I would propose to merge this PR right now and fix possible incoming Java 9 issues in separate PRs

bric3 commented:
Instead may we could detect the API level of ASM. Not sure what's the best way, to start we can detect if `Opcodes` contains `V1_9` or `ASM6`, see [branch of asm 6](http://websvn.ow2.org/filedetails.php?repname=asm&path=%2Fbranches%2FASM_6_FUTURE%2Fasm%2Fsrc%2Forg%2Fobjectweb%2Fasm%2FOpcodes.java). Bytebuddy repackages it in `net.bytebuddy.jae.asm.Opcodes`

bric3 commented:
Also shall we update jacoco [snapshot](http://www.eclemma.org/jacoco/trunk/doc/changes.html), there's a new build

TimvdLippe commented:
> Also shall we update jacoco snapshot, there's a new build

I would rather wait for the official 0.7.8 release and upgrade to that.

bric3 commented:
ok fair enough ;)

TimvdLippe commented:
ShallI merge this one?

bric3 commented:
Yes
